aten/src/ATen/record_function.cpp

```diff
@@ -428,4 +428,55 @@ void RecordFunction::end() {
   active_ = false;
 }
 
+//
+namespace {
+std::atomic<int> global_record_all_functions_ {0};
+thread_local int tls_record_all_functions_ = 0;
+thread_local int dispatch_tries_left_ = 0;
+double extra_sampling_prob_ = kLowProb;
+}
+void setRecordAllFunctionsGlobal() {
+  ++global_record_all_functions_;
+}
+void unsetRecordAllFunctionsGlobal() {
+  TORCH_CHECK(--global_record_all_functions_ >= 0);
+}
+void setRecordAllFunctionsLocal() {
+  ++tls_record_all_functions_;
+}
+void unsetRecordAllFunctionsLocal() {
+  TORCH_CHECK(--tls_record_all_functions_ >= 0);
+}
+
+void setExtraSamplingProbability(double sampling_prob) {
+  TORCH_CHECK(sampling_prob >= 0.0 && sampling_prob <= 1.0,
+          "Invalid sampling probability");
+  extra_sampling_prob_ = sampling_prob;
+}
+
+bool shouldRunRecordFunction() {
+  if (global_record_all_functions_ > 0) {
```

Comment:
if we don't care about a little slop, this can be `global_record_all_functions_.load(std::memory_order_relaxed)` and will perform better on non-x86 architectures. similarly, I think the increment/decrement can use relaxed as well in that case.